src/allmydata/interfaces.py

```diff
@@ -3111,3 +3120,199 @@ def get_client_endpoint():
         """
         Return an ``IStreamClientEndpoint``.
         """
+
+
+@attr.s
+class ImmutableCreateResult(object):
+    """Result of creating a storage index for an immutable."""
+    already_have = attr.ib(type=Set[int])
+    allocated = attr.ib(type=Set[int])
+
+
+@attr.s
+class WriteVector(object):
+    """Data to write to a chunk."""
+    offset = attr.ib(type=int)
+    data = attr.ib(type=bytes)
+
+
+class TestVectorOperator(Enum):
+    """Possible operators for test vectors."""
+    LT = b"lt"
+    LE = b"le"
+    EQ = b"eq"
+    NE = b"ne"
+    GE = b"ge"
+    GT = b"gt"
+
+
+@attr.s
+class TestVector(object):
+    """Checks to make on a chunk before writing to it."""
+    offset = attr.ib(type=int)
+    size = attr.ib(type=int)
+    operator = attr.ib(type=TestVectorOperator)
+    specimen = attr.ib(type=bytes)
+
+
+@attr.s
+class ReadVector(object):
+    """Reads to do on chunks operated on by the test/write vectors."""
+    offset = attr.ib(type=int)
+    size = attr.ib(type=int)
+
+
+@attr.s
+class TestWriteVectors(object):
+    """Test and write vectors for a specific share."""
+    test_vectors = attr.ib(type=List[TestVector])
+    write_vectors = attr.ib(type=List[WriteVector])
+    new_length = attr.ib(type=int)
+
+
+@attr.s
+class ReadTestWriteResult(object):
+    """Result of sending read-test-write vectors."""
+    success = attr.ib(type=bool)
+    # Map share numbers to reads corresponding to the request's list of
+    # ReadVectors:
+    reads = attr.ib(type=Dict[int, List[bytes]])
+
+
+class IStorageClientV2(Interface):
+    """
+    An improved interface for talking to the storage client, specifically for
+    immutable interactions.
```

Comment:
It looks like this grew support for mutable interactions too.  I wouldn't mind if mutable were separate from immutable were separate from .. lease and version, I guess?